components/remerge/src/vclock.rs

```diff
@@ -0,0 +1,162 @@
+            if sv > ov {
+                seen_gt = true;
+            }
+            if sv < ov {
+                seen_lt = true;
+            }
+            if seen_gt && seen_lt {
+                // No need to keep going once we've seen both.
+                return ClockOrdering::Conflicting;
+            }
+        }
+        match (seen_gt, seen_lt) {
+            (false, false) => ClockOrdering::Equivalent,
+            (true, false) => ClockOrdering::Ancestor,
+            (false, true) => ClockOrdering::Descendent,
+            (true, true) => ClockOrdering::Conflicting,
+        }
+    }
+
+    pub fn are_equivalent(&self, o: &VClock) -> bool {
+        self.get_ordering(o) == ClockOrdering::Equivalent
+    }
+
+    pub fn is_ancestor_of(&self, o: &VClock) -> bool {
+        self.get_ordering(o) == ClockOrdering::Ancestor
+    }
+
+    pub fn is_descendent_of(&self, o: &VClock) -> bool {
+        self.get_ordering(o) == ClockOrdering::Descendent
+    }
+
+    pub fn are_conflicting(&self, o: &VClock) -> bool {
+        self.get_ordering(o) == ClockOrdering::Conflicting
+    }
+
+    /// Get the clock's value for client_id, or 0 if it hasn't seen it.
+    pub fn get(&self, client_id: &Guid) -> Counter {
+        self.0.get(&client_id).copied().unwrap_or_default()
+    }
+
+    /// Add one to the clock's value for client_id
+    pub fn increment(&mut self, client_id: Guid) {
+        *self.0.entry(client_id).or_default() += 1
+    }
+
+    /// Assign `value` for client_id directly. Usually you want `apply` instead
+    pub fn set_directly(&mut self, client_id: Guid, value: Counter) {
+        if value == 0 {
+            self.0.remove(&client_id);
+        } else {
+            self.0.insert(client_id, value);
+        }
+    }
+
+    /// If `value` is greater than the current value for client_id store that
+    /// instead. Otherwise, do nothing.
+    ///
+    /// Notes that this clock has seen the `value`th event of `client_id`.
+    pub fn apply(&mut self, client_id: Guid, value: Counter) {
+        if value == 0 {
+            // Avoid inserting 0 if we can help it.
+            return;
+        }
+        let old_value = self.0.entry(client_id).or_default();
+        if *old_value < value {
+            *old_value = value;
+        }
+    }
+
+    pub fn apply_all(&mut self, o: &VClock) {
+        for (id, ctr) in &o.0 {
+            self.apply(id.clone(), *ctr)
+        }
+    }
+
+    pub fn merge(&self, o: &VClock) -> Self {
+        let mut res = self.clone();
+        res.apply_all(o);
+        res
+    }
+}
+
+impl PartialOrd for VClock {
+    fn partial_cmp(&self, other: &VClock) -> Option<std::cmp::Ordering> {
+        match self.get_ordering(other) {
+            ClockOrdering::Equivalent => Some(std::cmp::Ordering::Equal),
+            ClockOrdering::Ancestor => Some(std::cmp::Ordering::Greater),
+            ClockOrdering::Descendent => Some(std::cmp::Ordering::Less),
```

Comment:
Isn't an ancestor earlier and thus lesser? I'm having a hard time keeping the polarities straight here.